flink-tests/src/test/java/org/apache/flink/test/checkpointing/IgnoreInFlightDataITCase.java

```diff
@@ -98,75 +123,122 @@ public void testIgnoreInFlightDataDuringRecovery() throws Exception {
+        private boolean isRunning = true;
+
+        public NumberSource(SharedReference<AtomicInteger> lastCheckpointValue) {
+            this.lastCheckpointValue = lastCheckpointValue;
+        }
 
         @Override
         public void run(SourceContext<Integer> ctx) throws Exception {
             Iterator<Integer> stateIt = valueState.get().iterator();
             boolean isRecovered = stateIt.hasNext();
 
             if (isRecovered) {
-                Integer lastValue = stateIt.next();
+                synchronized (ctx.getCheckpointLock()) {
+                    Integer lastValue = stateIt.next();
 
-                // Checking that ListState is recovered correctly.
-                assertEquals(lastCheckpointedValue, lastValue.intValue());
+                    // Checking that ListState is recovered correctly.
+                    assertEquals(lastCheckpointValue.get().intValue(), lastValue.intValue());
 
-                // if it is started after recovery, just send one more value and finish.
-                ctx.collect(lastValue + 1);
+                    // if it is started after recovery, just send one more value and finish.
+                    ctx.collect(lastValue + 1);
+                }
             } else {
                 int next = 0;
-                while (true) {
+                while (isRunning) {
                     synchronized (ctx.getCheckpointLock()) {
-                        next++;
-                        valueState.update(singletonList(next));
-                        ctx.collect(next);
+                        // Emit data by batches to reduce the probability that before the first
+                        // checkpoint will be generated not enough data.
+                        do {
+                            next++;
+                            valueState.update(singletonList(next));
+                            ctx.collect(next);
+                        } while (next % 50 != 0 && isRunning);
                     }
+
+                    // Avoid the huge backpressure.
+                    LockSupport.parkNanos(100000);
                 }
             }
         }
 
         @Override
-        public void cancel() {}
+        public void cancel() {
+            isRunning = false;
+        }
 
         @Override
         public void snapshotState(FunctionSnapshotContext context) throws Exception {
-            if (lastCheckpointedValue > 0) {
+            if (lastCheckpointValue.get().get() > 0) {
                 throw new RuntimeException("Error during snapshot");
             }
 
-            lastCheckpointedValue = valueState.get().iterator().next();
+            Integer state = valueState.get().iterator().next();
+
+            if (state < PARALLELISM) {
+                // Try to restart task.
+                throw new RuntimeException(
+                        "Not enough data to guarantee the in-flight data were generated before the first checkpoint");
```

Comment:
This will fail with NoSuchElementException (the effect is the same, but less obvious).
And I think the allowed number of restarts should be increased (maybe infinite - it depends on how soon the source will start producing records vs checkpoint is triggered).

Reply:
Unfortunately, it doesn't work as I expected. After this exception is thrown on the next attemp we have the checkpoint with number 2 but we have the configuration that we should ignore in-flight data only for number 1. In fact, it is exactly why you have the error when you have a delay before emitting. I need to think about how to resolve this problem.